File: python/consumer.py
#!/usr/bin/env python3
"""使用 Python SDK 订阅并解析 VDM JSON 或 Protobuf Payload。"""

from __future__ import annotations

import json
import os
import threading
import time

from vdm_mqtt_sdk import (
    DecodedPayload,
    VdmMqttClient,
    VdmMqttClientConfig,
    VdmTopics,
)


REQUIRED_SUFFIXES = {
    "telemetry",
    "attributes",
    "event",
    "3A",
    "evidence",
    "rpc/req",
    "rpc/resp",
    "image",
}


def setting(name: str, default: str) -> str:
    return os.getenv(name, default)


class Consumer:
    def __init__(self) -> None:
        self.profile = setting("VDM_PAYLOAD_FORMAT", "protobuf")
        self.run_id = setting("VDM_RUN_ID", "manual")
        self.name = setting("VDM_CONSUMER_NAME", "python")
        self.timeout = float(setting("VDM_WAIT_TIMEOUT", "45"))
        self.seen: set[str] = set()
        self.rpc_started = False
        self.rpc_passed = False
        self.finished = False
        self.lock = threading.Lock()
        self.client = VdmMqttClient(
            VdmMqttClientConfig(
                host=setting("MQTT_HOST", "127.0.0.1"),
                port=int(setting("MQTT_PORT", "1883")),
                topics=VdmTopics(setting("VDM_BASE_TOPIC", "vdm/DEMO001")),
                payload_format=self.profile,
                connect_timeout=self.timeout,
                client_id=f"vdm-example-{self.name}-{self.run_id}",
            ),
            on_message=self.on_message,
            on_error=self.fail,
        )

    def control_topic(self, kind: str) -> str:
        return f"vdm-example/{self.run_id}/{kind}/{self.name}"

    def publish_control(self, kind: str, value: str) -> None:
        self.client.publish_raw(self.control_topic(kind), value, retain=True)

    def fail(self, reason: Exception | str) -> None:
        with self.lock:
            if self.finished:
                return
            self.finished = True
        print(f"FAIL consumer={self.name} reason={reason}", flush=True)
        try:
            self.publish_control("result", f"FAIL:{reason}")
        except Exception as exc:
            print(f"FAIL consumer={self.name} 无法上报结果: {exc}", flush=True)

    def on_message(self, message: DecodedPayload) -> None:
        if message.suffix not in REQUIRED_SUFFIXES:
            self.fail(f"未支持的 Topic: {message.suffix}")
            return
        print(
            f"DECODED consumer={self.name} profile={self.profile} "
            f"topic={message.suffix} bytes={len(message.raw)} "
            f"data={json.dumps(message.as_dict(), ensure_ascii=False, separators=(',', ':'))}",
            flush=True,
        )
        start_rpc = False
        with self.lock:
            if not self.finished:
                self.seen.add(message.suffix)
                if message.suffix == "telemetry" and not self.rpc_started:
                    self.rpc_started = True
                    start_rpc = True
        if start_rpc:
            threading.Thread(target=self.test_rpc, daemon=True).start()
        self.complete_if_ready()

    def test_rpc(self) -> None:
        try:
            response = self.client.call(
                "getAttr",
                {"keys": ["deviceId", "fwVer"]},
                req_id=11_001,
                timeout=10,
            )
            data = response.as_dict()
            if self.profile == "protobuf":
                device_id = data.get("getAttr", {}).get("attributes", {}).get("deviceId")
            else:
                device_id = data.get("data", {}).get("deviceId")
            if device_id != "DEMO001":
                raise AssertionError(f"RPC getAttr.deviceId 不匹配: {device_id}")
            with self.lock:
                self.rpc_passed = True
            print(
                f"RPC_PASS consumer={self.name} method=getAttr req_id=11001 data={data}",
                flush=True,
            )
            self.complete_if_ready()
        except Exception as exc:
            self.fail(exc)

    def complete_if_ready(self) -> None:
        with self.lock:
            should_pass = (
                not self.finished
                and self.rpc_passed
                and self.seen == REQUIRED_SUFFIXES
            )
            if should_pass:
                self.finished = True
        if should_pass:
            self.publish_control("result", "PASS")
            print(
                f"PASS consumer={self.name} profile={self.profile} "
                f"topics={','.join(sorted(self.seen))} rpc=getAttr",
                flush=True,
            )

    def run(self) -> None:
        self.client.start()
        self.publish_control("ready", "READY")
        print(f"READY consumer={self.name} profile={self.profile}", flush=True)
        while True:
            time.sleep(3600)


if __name__ == "__main__":
    Consumer().run()

File: javascript/consumer.js
#!/usr/bin/env node
"use strict";

const {
  VdmMqttClient,
  VdmTopics,
  parsePayloadFormat,
} = require("./sdk");

const REQUIRED_SUFFIXES = new Set([
  "telemetry",
  "attributes",
  "event",
  "3A",
  "evidence",
  "rpc/req",
  "rpc/resp",
  "image",
]);

function setting(name, fallback) {
  return process.env[name] || fallback;
}

async function main() {
  const format = parsePayloadFormat(setting("VDM_PAYLOAD_FORMAT", "protobuf"));
  const runId = setting("VDM_RUN_ID", "manual");
  const name = setting("VDM_CONSUMER_NAME", "javascript");
  const topics = new VdmTopics(setting("VDM_BASE_TOPIC", "vdm/DEMO001"));
  const seen = new Set();
  let rpcStarted = false;
  let rpcPassed = false;
  let resultSent = false;
  let client;

  const controlTopic = (kind) => `vdm-example/${runId}/${kind}/${name}`;
  const publishResult = async (value) => {
    await client.publishRaw(controlTopic("result"), value, true);
  };
  const fail = async (error) => {
    if (resultSent) return;
    resultSent = true;
    const reason = error?.message ?? String(error);
    console.error(`FAIL consumer=${name} reason=${reason}`);
    try {
      await publishResult(`FAIL:${reason}`);
    } catch (publishError) {
      console.error(`FAIL consumer=${name} 无法上报结果: ${publishError.message}`);
    }
  };
  const completeIfReady = async () => {
    if (
      !resultSent
      && rpcPassed
      && [...REQUIRED_SUFFIXES].every((suffix) => seen.has(suffix))
    ) {
      resultSent = true;
      await publishResult("PASS");
      console.log(
        `PASS consumer=${name} profile=${format} `
        + `topics=${[...seen].sort().join(",")} rpc=getAttr`,
      );
    }
  };
  const testRpc = async () => {
    try {
      const response = await client.call(
        "getAttr",
        { keys: ["deviceId", "fwVer"] },
        { reqId: 14_001, timeoutMs: 10_000 },
      );
      const deviceId = format === "protobuf"
        ? response.data?.getAttr?.attributes?.deviceId
        : response.data?.data?.deviceId;
      if (deviceId !== "DEMO001") {
        throw new Error(`RPC getAttr.deviceId 不匹配: ${deviceId}`);
      }
      rpcPassed = true;
      console.log(
        `RPC_PASS consumer=${name} method=getAttr req_id=14001 `
        + `data=${JSON.stringify(response.data)}`,
      );
      await completeIfReady();
    } catch (error) {
      await fail(error);
    }
  };

  client = new VdmMqttClient(
    {
      host: setting("MQTT_HOST", "127.0.0.1"),
      port: Number(setting("MQTT_PORT", "1883")),
      topics,
      payloadFormat: format,
      clientId: `vdm-example-${name}-${runId}`,
      qos: 1,
      connectTimeoutMs: Number(setting("VDM_WAIT_TIMEOUT", "45")) * 1000,
    },
    (message) => {
      if (!REQUIRED_SUFFIXES.has(message.suffix)) {
        void fail(new Error(`未支持的 Topic: ${message.suffix}`));
        return;
      }
      console.log(
        `DECODED consumer=${name} profile=${format} topic=${message.suffix} `
        + `bytes=${message.raw.length} data=${JSON.stringify(message.data)}`,
      );
      seen.add(message.suffix);
      if (message.suffix === "telemetry" && !rpcStarted) {
        rpcStarted = true;
        void testRpc();
      }
      void completeIfReady();
    },
    (error) => void fail(error),
  );

  await client.start();
  await client.publishRaw(controlTopic("ready"), "READY", true);
  console.log(`READY consumer=${name} profile=${format}`);
}

main().catch((error) => {
  console.error(`FAIL consumer=javascript reason=${error.message}`);
  process.exitCode = 1;
});
